packages/common/nbstore/src/sync/doc/peer.ts

```diff
@@ -0,0 +1,631 @@
+
+  constructor(
+    readonly local: DocStorage,
+    readonly syncMetadata: SyncStorage,
+    readonly remote: DocStorage,
+    readonly options: DocSyncPeerOptions = {}
+  ) {}
+
+  private status: Status = {
+    docs: new Set<string>(),
+    connectedDocs: new Set<string>(),
+    jobDocQueue: new AsyncPriorityQueue(),
+    jobMap: new Map(),
+    remoteClocks: new ClockMap(new Map()),
+    pulledRemoteClocks: new ClockMap(new Map()),
+    pushedClocks: new ClockMap(new Map()),
+    syncing: false,
+    retrying: false,
+    errorMessage: null,
+  };
+  private readonly statusUpdatedSubject$ = new Subject<string | true>();
+
+  private readonly jobs = createJobErrorCatcher({
+    connect: async (docId: string, signal?: AbortSignal) => {
+      const pushedClock = this.status.pushedClocks.get(docId);
+      const clock = await this.local.getDocTimestamp(docId);
+
+      throwIfAborted(signal);
+      if (pushedClock === null || pushedClock !== clock?.timestamp) {
+        await this.jobs.pullAndPush(docId, signal);
+      } else {
+        const pulled = this.status.pulledRemoteClocks.get(docId);
+        if (pulled === null || pulled !== this.status.remoteClocks.get(docId)) {
+          await this.jobs.pull(docId, signal);
+        }
+      }
+
+      this.status.connectedDocs.add(docId);
+      this.statusUpdatedSubject$.next(docId);
+    },
+    push: async (
+      docId: string,
+      jobs: (Job & { type: 'push' })[],
+      signal?: AbortSignal
+    ) => {
+      if (this.status.connectedDocs.has(docId)) {
+        const maxClock = jobs.reduce(
+          (a, b) => (a.getTime() > b.clock.getTime() ? a : b.clock),
+          new Date(0)
+        );
+        const merged = await this.mergeUpdates(
+          jobs.map(j => j.update).filter(update => !isEmptyUpdate(update))
+        );
+        if (!isEmptyUpdate(merged)) {
+          const { timestamp } = await this.remote.pushDocUpdate(
+            {
+              docId,
+              bin: merged,
+            },
+            this.uniqueId
+          );
+          this.schedule({
+            type: 'save',
+            docId,
+            serverClock: timestamp,
+          });
+        }
+        throwIfAborted(signal);
+        await this.actions.updatePushedClock(docId, maxClock);
+      }
+    },
+    pullAndPush: async (docId: string, signal?: AbortSignal) => {
+      const docRecord = await this.local.getDoc(docId);
+
+      const stateVector =
+        docRecord && !isEmptyUpdate(docRecord.bin)
+          ? encodeStateVectorFromUpdate(docRecord.bin)
+          : new Uint8Array();
+      const remoteDocRecord = await this.remote.getDocDiff(docId, stateVector);
+
+      if (remoteDocRecord) {
+        const {
+          missing: newData,
+          state: serverStateVector,
+          timestamp: serverClock,
+        } = remoteDocRecord;
+        this.schedule({
+          type: 'save',
+          docId,
+          serverClock,
+        });
+        throwIfAborted(signal);
```

Comment:
why do we listen on abort again and again?

Reply:
To ensure that no further operations after the abort,
add `throwIfAborted` after each await